### DISCUSSION TITLE
GrainTracker with EBSDReader and creating new Grain ID causes error

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#22166] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,
Recently, I want to use EBSD data for the large-scale phase field simulation. However, I found that an error when `GrainTracker` automatically creates a new GrainID for some reason, which I think is probably due to the EBSD data file not being able to automatically create a new GrainID data or the `EBSDAvgData` type of data,

Here is part of the error message displayed to the terminal, and two input files,
```
[36m
*** Info ***
Using ./09_local_ebsd_misori_GBanisotropy/out_09_local_ebsd_misori_GBanisotropy_cp/0055 for recovery.[39m
[36mThe following total 2 aux variables:
  delta_theta
  kappa_op
are added for automatic output by MaterialOutputAction.[39m
Framework Information:
MOOSE Version:           git commit 9c2c6fe on 2022-03-23
LibMesh Version:         
PETSc Version:           3.16.5
SLEPc Version:           3.16.2
Current Time:            Thu Sep 22 20:09:56 2022
Executable Timestamp:    Thu Sep 22 20:09:39 2022

Parallelism:
  Num Processors:          112
  Num Threads:             1

Mesh: 
  Parallel Type:           distributed
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   
    Total:                 139183
    Local:                 420
    Min/Max/Avg:           353/2789/1242
  Elems:                   
    Total:                 124133
···
Time Step 56, time = 15.4223, dt = 0.417725
 0 Nonlinear |R| = [32m9.180162e-01[39m
 1 Nonlinear |R| = [32m5.768520e-03[39m
 2 Nonlinear |R| = [32m7.708989e-07[39m
 3 Nonlinear |R| = [32m5.226698e-11[39m
[32m Solve Converged![39m
  Finished Solving                                                                       [[33m 10.09 s[39m] [[33m   10 MB[39m]

Grain Tracker Status:
Grains active index 0: 75 -> 75
Grains active index 1: 51 -> 51
Grains active index 2: 44 -> 45++
Grains active index 3: 38 -> 38
Grains active index 4: 34 -> 34
Grains active index 5: 28 -> 28
Grains active index 6: 19 -> 19
Grains active index 7: 10 -> 10
···
[33mNucleating Grain Detected  (variable index: 0)
[39m[33mNucleating Grain Detected  (variable index: 2)
[39m[32mMarking Grain 182 as INACTIVE (variable index: 0)
[39m[33m
*****************************************************************************
Couldn't find a matching grain while working on variable index: 0
Creating new unique grain: 329
Grain ID: 329
Ghosted Entities: 
Local Entities: 
Halo Entities: 145153 145415 175509 175511 175551 175553 175555 175556 175559 175560 175561 175563 175564 175566 175567 175569 175570 175572 175573 175574 
Periodic Node IDs: 
BBoxes:
Max: (x,y,z)=( 122.867,  151.803,        0) Min: (x,y,z)=( 120.861,  149.799,        0)
Status:  MARKED
Orig IDs (rank, index): (100, 1) 
Var_index: 0
Min Entity ID: 175562
···
*****************************************************************************
[39mFinished inside of GrainTracker


[31m
*** ERROR ***
The following error occurred in the object "ebsd_reader", of type "EBSDReader".

Error! Index out of range in EBSDReader::indexFromIndex(), index: 0 size: 0[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1

[31m
*** ERROR ***
The following error occurred in the object "ebsd_reader", of type "EBSDReader".

Error! Index out of range in EBSDReader::indexFromIndex(), index: 254 size: 0[39m


[31m
*** ERROR ***
The following error occurred in the object "ebsd_reader", of type "EBSDReader".

Error! Index out of range in EBSDReader::indexFromIndex(), index: 11 size: 0[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
...
```
The detailed information is as follows,
- the message displayed to terminal: [slurm-7511327.txt](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error3/slurm-7511327.txt)
- the input file: [ebsd_local_misori.i](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error3/ebsd_local_misori.i)，
- the ebsd data file: [local_Ti700du_5minFill_refine_1.inl](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error3/local_Ti700du_5minFill_refine_1.inl)

Also, I referred to the advice of chaitanyaBhave26 in #20198, but it didn't work.

So, my question is how to get the EBSDReader to work during the calculation when the GrainTracker creates a new GrainID, instead of reporting `Error! Index out of range in EBSDReader::indexFromIndex(), index`.

Any suggestions or recommendations to fix these problems would be greatly appreciated.

Thanks

wei peng

**Comment by permcody** [comment, marked as answer]
Hi Wei, I don't have time to dive into this right now, but what I typically due to troubleshoot is to look at the solution (image) of the initial condition for the order parameter in question. Look to see if there are two grains that are exceptionally close (maybe just a few elements apart) that look like they may be close to contact. What can happen is when you start to solve the phase-field equations and those sharp interfaces start to diffuse, these grains can rapidly grow in just one time step. I haven't run into this very much (admittedly I've only dealt with a handful of real EBSD files so my sample size is also low).

The PolyCrystalInitialCondition system was designed with this scenario in mind. There is a "halo level" just for the IC that's baked into that class (I haven't exposed it as an input parameter, but you could change that value and recompile). The idea is that for these types of cases where we have two grains somewhat close to each other, we might be able to force the coloring algorithm to place them on different order parameters instead of placing them more closely by increasing that halo level (they'll intersect before being placed). 

Other solutions would be to try a different coloring algorithm just to see if you get lucky by perturbing the layout of how the grains are assigned to OPs. The back-tracking algorithm (although potentially the slowest of these - with potential exponential runtime!) typically does the best of spreading out your grains more evenly than the stochastic algorithms in PETSc.

Feel free to post images of your IC here and also "order parameter 2". I'd be happy to guide you through fixing this and will help as I can!

**Reply by PengWei97** [reply]
hi @permcody,
Thank you for your help. 

> What can happen is when you start to solve the phase-field equations and those sharp interfaces start to diffuse, these grains can rapidly grow in just one time step. 

You are right, the point where the new grain ID is created by GrainTracker is because the two grians are too cloase together. Then, these are my output images of gr0 and gr2,
![image](https://user-images.githubusercontent.com/56183775/191877179-31737f39-cc90-43a5-a317-3611e4a5cd58.png)
Based on the information displayed in the terminal by [slurm-7511327.txt](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error3/slurm-7511327.txt), I selected the two points where the GrainTracker acts, and the following is displayed,
![image](https://user-images.githubusercontent.com/56183775/191877469-d88cf5fc-394c-431b-91ba-827804d82aeb.png)

***
> Other solutions would be to try a different coloring algorithm just to see if you get lucky by perturbing the layout of how the grains are assigned to OPs.

Base on your suggestion, I simulated by using the bt algorithm, but unfortunately the same error was occured as well. Here is the relevant information and files,
```bash
[33mNucleating Grain Detected  (variable index: 3)
[39m[32mMarking Grain 182 as INACTIVE (variable index: 2)
[39m[33m
*****************************************************************************
Couldn't find a matching grain while working on variable index: 3
Creating new unique grain: 329
Grain ID: 329
Ghosted Entities: 
Local Entities: 
Halo Entities: 138015 138567 146670 146691 146693 146711 146712 147994 148023 148024 148025 148027 148028 148030 148031 148033 148034 148036 148037 148038 
Periodic Node IDs: 
BBoxes:
Max: (x,y,z)=( 122.867,  151.803,        0) Min: (x,y,z)=( 120.861,  149.799,        0)
Status:  MARKED
Orig IDs (rank, index): (87, 7) 
Var_index: 3
Min Entity ID: 148026


*****************************************************************************
[39mFinished inside of GrainTracker
```
- the output to the terminal: [slurm-7518068.txt](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error3/slurm-7518068.txt)
- the initial and last output exodus files: [10_local_ebsd_misori_GBanisotropy_bt](https://github.com/PengWei97/luwu/blob/gNGsGG2022/grainGrowth/gNGs2022/error3/)

***
>  The idea is that for these types of cases where we have two grains somewhat close to each other, we might be able to force the coloring algorithm to place them on different order parameters instead of placing them more closely by increasing that halo level (they'll intersect before being placed).

Next, I will try to modify the 'halo level' in `GrainTracker` or `FeatureFloodCount`. Please forgive me for not being very familiar with this piece, but I will do my best to understand it. Of course it would be nice if you facilitate the area where the code is provided.

wei

**Reply by PengWei97** [reply]
Hi @permcody, 

This problem seems to be solved by putting `max_h_level = 3` in the `Adaptivity` module. Also next, I will read carefully your two papers and thesis about this algorithm.

Wei

**Reply by permcody** [reply]
While you generally do want to set a `max_h_level` (depending on your refine and coarsen fractions), I wouldn't expect this to be the solution to this problem! I'm glad you have a workaround nonetheless though. Let me know if I can answer any further questions on the subject.

**Reply by PengWei97** [reply]
Yeah, you are right. This method can only temporarily solve the problem, and who knows when the same problem will occur again. So, when I get the issue I submitted yesterday resolved about `GBAnisotropy`, I will continue to try to fix the problem by turning off the ability to create new grain IDs and making the remapping work only in `GrainTracker`.
